### PR TITLE
Add material laws to the HT process

### PR DESCRIPTION
**Pull request by TomFischer**
The following material properties are now configurable via the project file:
- fluid density
- viscosity
- porosity
- permeability
- storage

ToDo:
- [x] doxygen documentation

### CODE REVIEW

File: ProcessLib/HT/CreateHTProcess.cpp [hunk 1]
```diff
@@ -11,10 +11,15 @@
...
 #include "HTProcessData.h"
+
 #include "ProcessLib/Parameter/ConstantParameter.h"
 #include "ProcessLib/Utils/ParseSecondaryVariables.h"
 #include "ProcessLib/Utils/ProcessUtils.h"
 
+#include "CreatePorousMediaProperties.h"
+#include "MaterialLib/Fluid/Density/createFluidDensityModel.h"
+#include "MaterialLib/Fluid/Viscosity/createViscosityModel.h"
+
```
Comment (endJunction) [inline ProcessLib/HT/CreateHTProcess.cpp]: Could you reorder the includes please:
CreateHTProcess.h first (as it is) then
ProcessLib/* MaterialLib/* (probably alphabetic order) then
HTProcess.h, HTProcessData.h etc from local directory.

File: ProcessLib/HT/CreateHTProcess.cpp [hunk 2]
```diff
@@ -45,45 +50,18 @@ std::unique_ptr<Process> createHTProcess(
...
-
-    // Parameter for the viscosity.
-    auto& viscosity = findParameter<double>(
-        config,
-        //! \ogs_file_param_special{prj__processes__process__HT__viscosity}
-        "viscosity", parameters, 1);
-    DBUG("Use \'%s\' as viscosity parameter.", viscosity.name.c_str());
+    //! \ogs_file_param_special{prj__processes__process__HT__fluid}
+    auto const& fluid_config = config.getConfigSubtree("fluid");
+    //! \ogs_file_param_special{prj__processes__process__HT__fluid__viscosity}
```
Comment (endJunction) [inline ProcessLib/HT/CreateHTProcess.cpp]: ~~special~~ 
This is an ordinary `\ogs_file_param`.

File: ProcessLib/HT/CreateHTProcess.cpp [hunk 3]
```diff
@@ -92,12 +70,18 @@ std::unique_ptr<Process> createHTProcess(
...
+
     // Parameter for the density of the fluid.
-    auto& density_fluid = findParameter<double>(
+    auto& fluid_reference_density= findParameter<double>(
         config,
-        //! \ogs_file_param_special{prj__processes__process__HT__density_fluid}
-        "density_fluid", parameters, 1);
-    DBUG("Use \'%s\' as density_fluid parameter.", density_fluid.name.c_str());
+        //! \ogs_file_param_special{prj__processes__process__HT__fluid_reference_density}
+        "fluid_reference_density", parameters, 1);
```
Comment (endJunction) [inline ProcessLib/HT/CreateHTProcess.cpp]: Not sure, but maybe it should be `prj__processes__process__HT__fluid__reference_density` (extra underscore after fluid)?

File: ProcessLib/HT/CreatePorousMediaProperties.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    if (mesh_material_ids)
+    {
+        material_ids.reserve(mesh_material_ids->size());
+        std::copy(mesh_material_ids->cbegin(), mesh_material_ids->cend(),
+                  material_ids.begin());
+    }
+    else
+    {
+        material_ids.resize(mesh.getNumberOfElements());
+        std::fill_n(material_ids.begin(), mesh.getNumberOfElements(), 0);
```
Comment (endJunction) [inline ProcessLib/HT/CreatePorousMediaProperties.cpp]: fill after resize is redundant. If you want to be sure, pass a required value to resize as second argument.
http://en.cppreference.com/w/cpp/container/vector/resize

File: ProcessLib/HT/HTFEM.h [hunk 5]
```diff
@@ -21,14 +21,26 @@
...
+struct IntegrationPointData final
+{
+    IntegrationPointData(NodalRowVectorType N_, GlobalDimNodalMatrixType dNdx_,
+                         double integration_weight_)
+        : N(N_), dNdx(dNdx_), integration_weight(integration_weight_)
+    {}
+
+    NodalRowVectorType N;
+    GlobalDimNodalMatrixType dNdx;
+    double integration_weight;
```
Comment (endJunction) [inline ProcessLib/HT/HTFEM.h]: const (and maybe in ctor's arg list too :)

File: ProcessLib/HT/PorousMediaProperties.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    MaterialLib::PorousMedium::Storage const& getSpecificStorage(
+        double t, SpatialPosition const& pos) const;
+
+private:
+    std::vector<std::unique_ptr<MaterialLib::PorousMedium::Porosity>>
+        _porosity_models;
+    std::vector<Eigen::MatrixXd> _intrinsic_permeability_models;
+    std::vector<std::unique_ptr<MaterialLib::PorousMedium::Storage>>
+        _specific_storage_models;
+    int getMaterialID(SpatialPosition const& pos) const;
```
Comment (endJunction) [inline ProcessLib/HT/PorousMediaProperties.h]: Please separate function from variable declarations. I'd do it like this (but choose something different if you like):
```c++
class A
{
public:
  void f();
private:
  int g();
private:
  double _var;
```

File: ProcessLib/HT/CreatePorousMediaProperties.cpp (hunk 4 above)
Comment (endJunction) [inline ProcessLib/HT/CreatePorousMediaProperties.cpp]: Alternatively you could use boost::optional like here: https://github.com/endJunction/ogs/commit/7c4fdc825d240094e292279ef9791c32bb9f2a88

File: ProcessLib/HT/CreateHTProcess.cpp [hunk 7]
```diff
@@ -78,7 +78,7 @@ std::unique_ptr<Process> createHTProcess(
     // Parameter for the density of the fluid.
     auto& fluid_reference_density= findParameter<double>(
         config,
-        //! \ogs_file_param_special{prj__processes__process__HT__fluid_reference_density}
+        //! \ogs_file_param{prj__processes__process__HT__fluid_reference_density}
```
Comment (endJunction) [inline ProcessLib/HT/CreateHTProcess.cpp]: Same for other getConfigXXX() ;)

File: ProcessLib/HT/HTProcessData.h [hunk 8]
```diff
@@ -91,20 +89,17 @@ struct HTProcessData
     //! Assignments are not needed.
     void operator=(HTProcessData&&) = delete;
 
-    Parameter<double> const& porosity;
-    Parameter<double> const& intrinsic_permeability;
-    Parameter<double> const& specific_storage;
-    Parameter<double> const& viscosity;
+    PorousMediaProperties porous_media_properties;
+    std::unique_ptr<MaterialLib::Fluid::FluidProperty> viscosity_model;
```
Comment (wenqing) [inline ProcessLib/HT/HTProcessData.h]: You can replace
```
std::unique_ptr<MaterialLib::Fluid::FluidProperty> viscosity_model;
std::unique_ptr<MaterialLib::Fluid::FluidProperty> fluid_density;
Parameter<double> const& specific_heat_capacity_fluid;
Parameter<double> const& thermal_conductivity_fluid;
```
with 
```
const std::unique_ptr<MaterialLib::Fluid::FluidProperties>  _fluid_properties;
```
There is a creator for ``MaterialLib::Fluid::FluidProperties`` as
```
std::unique_ptr<FluidProperties> createFluidProperties(
    BaseLib::ConfigTree const& config);
```

File: ProcessLib/HT/CreateHTProcess.cpp [hunk 9]
```diff
@@ -92,12 +70,18 @@ std::unique_ptr<Process> createHTProcess(
         "density_solid", parameters, 1);
     DBUG("Use \'%s\' as density_solid parameter.", density_solid.name.c_str());
 
+    //! \ogs_file_param_special{prj__processes__process__HT__fluid__density}
+    auto const& fluid_density_conf = fluid_config.getConfigSubtree("density");
```
Comment (wenqing) [inline ProcessLib/HT/CreateHTProcess.cpp]: Fluid density, viscosity, heat capacity and thermal conductivity can be represented by FluidProperties and its instance can be created by calling created by createFluidProperties(BaseLib::ConfigTree const& config) here. (see  LiquidFlowMaterialProperties.h/cpp).

File: ProcessLib/HT/CreateHTProcess.cpp (hunk 3 above)
Comment (TomFischer) [inline ProcessLib/HT/CreateHTProcess.cpp]: I think the code is correct at this point.

File: ProcessLib/HT/CreatePorousMediaProperties.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    BaseLib::reorderVector(porosity_models, mat_ids);
+    BaseLib::reorderVector(storage_models, mat_ids);
+
+    std::vector<int> material_ids;
+    auto const& mesh_material_ids =
+        mesh.getProperties().getPropertyVector<int>("MaterialIDs");
+    if (mesh_material_ids)
+    {
+        material_ids.reserve(mesh_material_ids->size());
+        std::copy(mesh_material_ids->cbegin(), mesh_material_ids->cend(),
```
Comment (TomFischer) [inline ProcessLib/HT/CreatePorousMediaProperties.cpp]: Yes, that is true that after copying the values of `mesh_material_ids` it is not used any longer. But it is a reference to a mesh internal `PropertyVector`. Releasing the memory would destroy this mesh internal `PropertyVector`.

**Pull request merged**